<html>
  <head>
    <title>RNAstructure Text Interfaces Help: DuplexFold</title>
    <link
       href="Stylesheet.css"
       rel="stylesheet"
       type="text/css"
    />
  </head>
  <body>
    <div class="bodydiv">
      <table cellpadding="0" cellspacing="0" border="0" class="mainTable">
	<tr class="headerRow">
	  <td class="headerBox">
	    <img src="icon.jpg" alt="RNAstructure logo"/>
	  </td>
	  <td class="headerLabel">
	    <h1>
	      RNAstructure Help
	      <br/>
	      DuplexFold
	    </h1>
	  </td>
	  <td class="headerBox2">
	    <h3><ul>
	      <li><a href="index.html">Contents</a></li>
	      <li><a href="AlphabeticalIndex.html">Index</a></li>
	    </ul></h3>
	  </td>
	</tr>
	<tr><td class="mainContent" colspan="3">
	    <p>
	      DuplexFold is used to predict lowest free energy structures
	      containing two strands. The structures
              <span class="boldBlack">will not</span> contain intramolecular
              pairs.
	      <br/>
	      Note that output is written to a CT file where the sequences are
	      concatenated, with an intermolecular linker between them ("III").
	    </p>

	    <a name="Usage"><h3>
	      USAGE: DuplexFold &lt;seq file 1&gt; &lt;seq file 2&gt;
	      &lt;ct file&gt; [options]
	    </h3></a>

	    <h3 class="leftHeader">Required parameters:</h3>
	    <table>
	      <tr>
		<td class="cell">&lt;seq file 1&gt;</td>
		<td>The name of a file containing a first input sequence.Note that lowercase nucleotides are forced single-stranded in structure prediction.</td>
	      </tr>
              <tr>
                <td class="cell">&lt;seq file 2&gt;</td>
                <td>The name of a file containing a second input sequence. Note that lowercase nucleotides are forced single-stranded in structure prediction.</td>
              </tr>
	      <tr>
		<td class="cell">&lt;ct file&gt;</td>
		<td>The name of a CT file to which output will be written.</td>
	      </tr>
	    </table>

	    <h3 class="leftHeader">
	      Options which don't require added values:
	    </h3>
	    <table>
	      <tr>
		<td class="cell">-d, -D, --DNA</td>
		<td>
		  Specify that the sequence is DNA, and DNA parameters are to
		  be used. Note that the backbone type must be the same for
                  both sequences.
		  <br/>
		  Default is to use RNA parameters.
		</td>
	      </tr>
	      <tr>
		<td class="cell">-h, -H, --help</td>
		<td>Display the usage details message.</td>
	      </tr>
	    </table>

	    <h3 class="leftHeader">Options which require added values:</h3>
	    <table>
	      <tr>
		<td class="cell">-l, -L, --loop</td>
		<td>
		  Specify the maximum number of unpaired nucleotides in an internal or bulge loop.
              <br/>
                  Default is 6 unpaired nucleotides.
		</td>
	      </tr>
              <tr>
		<td class="cell">-m, -M, --maximum</td>
		<td>
                  Specify a maximum number of structures. Note that suboptimal
		  structures are generated until either the maximum number of
		  structures are reached or the maximum percent difference is
		  reached (below).
                  <br/>
                  Default is 10 structures.
                </td>
              </tr>
              <tr>
		<td class="cell">-p, -P, --percent</td>
		<td>
                  Specify a maximum percent difference in folding free energy
		  change in suboptimal structures. Note that suboptimal
		  structures are generated until either the maximum percent
		  free energy difference is reached or until the maximum number
		  of structures is reached (above).
		  <br/>
		  Default is 40 percent (specified as 40, not 0.4).
                </td>
              </tr>
	      <tr>
		<td class="cell">-t, -T, --temperature</td>
		<td>
		  Specify the temperature at which calculation takes place in
		  Kelvin.
		  <br/>
		  Default is 310.15 K, which is 37 degrees C.
		</td>
	      </tr>
	      <tr>
		<td class="cell">-w, -W, --window</td>
		<td>
		  Specify a window size for generating suboptimal structures.  Smaller windows (down to zero) will result in a larger set of structures that are similar.  Larger windows result in fewer structures that are more different in predicted pairs.
		  <br/>
		  Default is 0 nucleotides.
		</td>
	      </tr>
	    </table>

	    <a name="Refs"><h3 class="leftHeader">References</h3></a>
            <ol class="noListTopMargin">
              <li class="gapListItem">
                Reuter, J.S. and Mathews, D.H.
                <br/>
                "RNAstructure: software for RNA secondary structure prediction
                and analysis."
                <br/>
                <i>BMC Bioinformatics</i>, 11:129. (2010).
              </li>
              <li class="gapListItem">
                  Piekna-Przybylska, D., DiChiacchio, L., Mathews, D.H. and Bambara, R.A.
                  <br/>
                  "A sequence similar to tRNA3Lys gene is embedded in HIV-1 U3/R and promotes minus strand transfer."
                  <br/>
                  <i>Nature Structural & Molecular Biology</i>, 17:83-89. (2010).
              </li>
            </ol>

	</td></tr>
	<tr><td class="footer" colspan="3">
	  <p>
	    Visit
	    <a href="http://rna.urmc.rochester.edu/RNAstructure.html">
	      The Mathews Lab RNAstructure Page</a>
	    for updates and latest information.
	  </p>
	</td></tr>
      </table>
    </div>
  </body>
</html>
