<html>
  <head>
    <title>RNAstructure Text Interfaces Help: dynalign</title>
    <link
       href="Stylesheet.css"
       rel="stylesheet"
       type="text/css"
    />
  </head>
  <body>
    <div class="bodydiv">
      <table cellpadding="0" cellspacing="0" border="0" class="mainTable">
	<tr class="headerRow">
	  <td class="headerBox">
	    <img src="icon.jpg" alt="RNAstructure logo"/>
	  </td>
	  <td class="headerLabel">
	    <h1>
	      RNAstructure Help
	      <br/>
	      dynalign
	    </h1>
	  </td>
	  <td class="headerBox2">
	    <h3><ul>
	      <li><a href="index.html">Contents</a></li>
	      <li><a href="AlphabeticalIndex.html">Index</a></li>
	    </ul></h3>
	  </td>
	</tr>
	<tr><td class="mainContent" colspan="3">
	    <p>
	      dynalign is used to find the lowest free energy common secondary structures for two homologous sequences. It comes in two variants, dynalign and
           dynalign-smp. dynalign runs structure determination on a single processor,
           while dynalign-smp runs in parallel on a shared memory machine.
	    </p>

	    <a name="Usage"><h3>
	      USAGE: dynalign (or dynalign-smp) &lt;config file&gt;
	    </h3></a>

	    <h3 class="leftHeader">Required parameters:</h3>
	    <table>
	      <tr>
		<td class="cell">&lt;config file&gt;</td>
		<td>The name of a file containing required configuration data.</td>
	      </tr>
	    </table>

	    <h3 class="leftHeader">
	      Options which don't require added values:
	    </h3>
          <h4>NONE</h4>

	    <h3 class="leftHeader">Options which require added values:</h3>
	    <h4>NONE</h4>

         <h3 class="leftHeader">Config file format:</h3>
         <p>The following example is based on sample.conf, a standard example found in the dynalign directory of the RNAstructure repository.</p>

         <pre>
################################################################
# Required input
################################################################

inseq1 = &lt;seq file 1&gt;
inseq2 = &lt;seq file 2&gt;
outct = &lt;output ct file for seq 1&gt;
outct2 = &lt;output ct file for seq 2&gt;
aout = &lt;output alignment file&gt;

################################################################
# Options with default values if not explicitly specified
# (Default values are shown)
#################################################################

# fgap is the per nucleotide insert penalty for alignments:
fgap = .4

# maxtrace is the masximum number of predicted structures:
maxtrace = 750

# percent is the maximum % change in free energy from the lowest free energy structure:
percent = 20

# bpwin is the base pair window:
bpwin = 2

# awin is the alignment window:
awin = 1

# insert indicates whether single basepair inserts will be allowed:
insert = 1

# singlefold_subopt_percent is the maximum % difference in folding free energy change
# from single sequence folding for pairs that will be allowed in a subsequent Dynalign calculation.
# This is used to save calculation time by pre-screening allowed pairs. 
singlefold_subopt_percent = 30

# imaxseparation is the traditional M parameter:
# -99 indicates that the alignment constraint (preferred method is used)
imaxseparation = -99

# num_processor is required only for smp (parallel) calculations
num_processors = 1

# optimal only is optional, only the lowest free energy structure is calculated if optimal_only = 1
optimal_only = 0

# local folding is performed if local = 1, the default is 0 (global folding)
local = 0

# the following are needed for progressive calculations
# dsv_templated is set to 1 to read the template from previous calculation
dsv_templated = 0
dsvtemplatename = &lt;template file name&gt;

# The following are used to predict a structure for sequence 2, where the structure for sequence 1 is known.
# If ct_templated is set to 1, inseq1 must refer to a ct file, NOT a sequence file.
ct_templated = 0

# The following parameters are used when SHAPE data is utilized (see below).
# There is a set of parameters for each sequence.
shapeslope1 = 2.6
shapeintercept1 = -0.8
shapeslope2 = 2.6
shapeintercept2 = -0.8

################################################################
# Options that are not required and have no default values
################################################################

# Savefiles are optional and are needed for dot plots.
savefile = &lt;save file name&gt;

# Folding constraints can be input using constraint files:
constraint_1_file = &lt;constraint file for seq 1&gt;
constraint_2_file = &lt;constraint file for seq 2&gt;

# SHAPE data can be input using .shape files for either, neither, or both
# SHAPE is utilized using the pseudo free energy method of Deigan et al.
# PNAS 106:97
shape_1_file = &lt;SHAPE file for seq 1&gt;
shape_2_file = &lt;SHAPE file for seq 2&gt;

# Use constraint_align_file to enforce specific nucleotide alignments.
constraint_align_file = &lt;alignment constraints file&gt;

# Use maximumpairingdistance to limit the maximum distance between 
# paired nucleotides (where the final # indicates the sequence #).
# Note that this only works for sequence 1 if the calculation is not 
# cttemplated or dsvtemplated.
maximumpairingdistance1 = &lt;value for seq 1&gt;
maximumpairingdistance2 = &lt;value for seq 2&gt;
         </pre>

	    <a name="Refs"><h3 class="leftHeader">References:</h3></a>
            <ol class="noListTopMargin">
              <li class="gapListItem">
                Reuter, J.S. and Mathews, D.H.
                <br/>
                "RNAstructure: software for RNA secondary structure prediction
                and analysis."
                <br/>
                <i>BMC Bioinformatics</i>, 11:129. (2010).
              </li>
              <li class="gapListItem">
                Harmanci, A.O., Sharma, G. and Mathews, D.H.
                <br/>
                "Efficient Pairwise RNA Structure Prediction Using Probabilistic Alignment Constraints in Dynalign."
                <br/>
                <i>BMC Bioinformatics</i>, 8:130. (2007)
              </li>
              <li class="gapListItem">
                Uzilov, A.V., Keegan, J.M. and Mathews, D.H.
                <br/>
                "Detection of non-coding RNAs on the basis of predicted secondary structure formation free energy change."
                <br/>
                <i>BMC Bioinformatics</i>, 7:173. (2006).
              </li>
              <li class="gapListItem">
                Mathews, D.H.
                <br/>
                In Baxevanis, A. D., Davison, D. B., Page, R. D. M., Petsko, G. A., Stein, L. D. and Stormo, G. D. (eds.).
                <br/>
                <i>Current Protocols in Bioinformatics</i>.
                <br/>
                John Wiley and Sons, Inc., New York, pp. 12.14.11-12.14.11. (2005).
              </li>
              <li class="gapListItem">
                Mathews, D.H.
                <br/>
                "Predicting a set of minimal free energy RNA secondary structures common to two sequences."
                <br/>
                <i>Bioinformatics</i>, 21:2246-2253. (2005).
              </li>
              <li class="gapListItem">
                Mathews, D.H. and Turner, D.H.
                <br/>
                "Dynalign: An algorithm for finding the secondary structure common to two RNA sequences."
                <br/>
                <i>J. Mol. Biol.</i>, 317:191-203. (2002).
              </li>
            </ol>
	</td></tr>
        <tr><td class="footer" colspan="3">
          <p>
            Visit
            <a href="http://rna.urmc.rochester.edu/RNAstructure.html">
              The Mathews Lab RNAstructure Page</a>
            for updates and latest information.
          </p>
        </td></tr>
      </table>
    </div>
  </body>
</html>
