<html>
  <head>
    <title>RNAstructure Help -- References</title>
    <link
      href="Styles.css"
      rel="stylesheet"
      type="text/css"
    />
  </head>
  <body>
    <h1>RNAstructure</h1>
    <h2>Version 5.1</h2>
    <h2>Copyright 1996 - 2010</h2>
    <h2>by David H. Mathews</h2>
    <br/>

    <h3>References</h3>

    <p><b>Research benefitting from RNAstructure should cite:</b></p>
    <p>J.S. Reuter and D.H. Mathews.</p>
    <p>"RNAstructure: software for RNA secondary structure prediction and analysis."</p>
    <p><i>BMC Bioinformatics</i>, 11:129. (2010).</p>
    <br/>

    <p>The algorithms central to RNAstructure are based on the following sources. Research benefitting from these capabilities encapsulated within RNAstructure should cite the appropriate sources.</p>
    <br/>

    <p><b>Most current folding parameters:</b></p>
    <p>D.H. Mathews, M.D. Disney, J.L. Childs, S.J. Schroeder, M. Zuker, and D.H. Turner.</p>
    <p>"Incorporating Chemical Modification Constraints into a Dynamic Programming Algorithm for Prediction of RNA Secondary Structure."</p>
    <p><i>Proceedings of the National Academy of Sciences USA</i>, 101:7287-7292. (2004).</p>
    <br/>

    <p><b>Previous reference to the folding algorithm, with details on thermodynamic parameter derivation:</b></p>
    <p>D.H. Mathews, J. Sabina, M. Zuker, and D.H. Turner.</p>
    <p>"Expanded Sequence Dependence of Thermodynamic Parameters Improves Prediction of RNA Secondary Structure."</p>
    <p><i>Journal of Molecular Biology</i>, 288:911-940. (1999).</p>
    <br/>

    <p><b>Prediction of secondary structures at temperatures other than 37&#176C:</b></p>
    <p>Z.J. Lu, D.H. Turner, and D.H. Mathews.</p>
    <p>"A Set of Nearest Neighbor Parameters for Predicting the Enthalpy Change of RNA Secondary Structure Formation."</p>
    <p><i>Nucleic Acids Research</i>, 34(17):4912-24. (2006).</p>
    <br/>

    <p><b>Partition function for calculating base pair probabilities:</b></p>
    <p>D.H. Mathews.</p>
    <p>"Using an RNA Secondary Structure Partition Function to Determine Confidence in Base Pairs Predicted by Free Energy Minimization."</p>
    <p><i>RNA</i>, 10:1178-1190. (2004).</p>
    <br/>

    <p><b>Dynalign:</b></p>
    <p>D.H. Mathews.</p>
    <p>"Predicting a Set of Minimal Free Energy Secondary Structures Common to Two Sequences."</p>
    <p><i>Bioinformatics</i>, 21:2246-2253. (2005).</p>
    <br/>

    <p><b>Extensive use of Dynalign (for extensive acceleration of the algorithm):</b></p>
    <p>A. Harmanci, G. Sharna, and D.H. Mathews.</p>
    <p>"Efficient Pairwise RNA Structure Prediction Using Probabilistic Alignment Constraints in Dynalign."</p>
    <p><i>BMC Bioinformatics</i>, 8:130-150. (2007).</p>
    <br/>

    <p><b>Previous reference to Dynalign:</b></p>
    <p>D.H. Mathews and D.H. Turner.</p>
    <p>"Dynalign: An Algorithm for Finding the Secondary Structure Common to Two RNA Sequences."</p>
    <p><i>Journal of Molecular Biology</i>, 317:191-203. (2002).</p>
    <br/>

    <p><b>OligoScreen:</b></p>
    <p>D.V. Mateeva, D.H. Mathews, A.D. Tsodikov, S.A. Shabalina, R.F. Gesteland, J.F. Atkins, and S.M. Freier.</p>
    <p>"Thermodynamic Criteria for High Hit Rate Antisense Oligonucleotide Design."</p>
    <p><i>Nucleic Acids Research</i>, 31:4989-4994. (2003).</p>
    <br/>

    <p><b>OligoWalk:</b></p>
    <p>D.H. Mathews, M.E. Burkard, S.M. Freier, J.R. Wyatt, and D.H. Turner.</p>
    <p>"Predicting Oligonucleotide Affinity to Nucleic Acid Targets."</p>
    <p><i>RNA</i>, 5:1458-1469. (1999).</p>
    <br/>

    <p><b>DNA/RNA duplex parameters for DNA oligomers:</b></p>
    <p>N. Sugimoto, S. Nakano, A. Katoh, H. Nakamura, T. Ohmichi, M. Yoneyama, and M. Sasaki.</p>
    <p>"Thermodynamic Parameters to Predict the Stability of RNA/DNA Hybrid Duplexes."</p>
    <p><i>Biochemistry</i>, 34:11211-11216. (1995).</p>
    <br/>

    <p><b>Prediction of all suboptimal secondary structures:</b></p>
    <p>S. Dian, D.H. Mathews, and D.H. Turner.</p>
    <p>"Interpreting Oligonucleotide Microarray Data to Determine RNA Secondary Structure. Application to the 3' End of Bombyx mori R2 DNA."</p>
    <p><i>Biochemistry</i>, 45:9819-9832. (2006).</p>
    <br/>

    <p><b>Methods for predicting suboptimal secondary structures:</b></p>
    <p>D.H. Mathews.</p>
    <p>"Revolutions in RNA Secondary Structure Prediction."</p>
    <p><i>Journal of Molecular Biology</i>, 359:526-532. (2006).</p>
    <br/>

    <p><b>Original reference to suboptimal secondary structure prediction algorithm:</b></p>
    <p>S. Wuchty, W. Fontant, I,L. Hofacker, and P. Schuster.</p>
    <p>"Complete Suboptimal Folding of RNA and the Stability of Secondary Structures."</p>
    <p><i>Biopolymers</i>, 49:145-165. (1999).</p>
    <br/>

    <p><b>Stochastic sampling:</b></p>
    <p>Y. Ding and C.E. Lawrence.</p>
    <p>"A Statistical Sampling Algorithm for RNA Secondary Structure Prediction."</p>
    <p><i>Nucleic Acids Research</i>, 31:7280-7301. (2003).</p>
    <br/>

    <p><b>Using SHAPE data to constrain structure prediction</b></p>
    <p>K.A. Wilkinson, R.J. Gorelick, S.M. Vasa, N. Guex, A. Rein, D.H. Mathews, M.C. Giddings, K.M. Weeks.</p>
    <p>"High-Throughput SHAPE Analysis Reveals Structures in HIV-1 Genomic RNA Strongly Conserved Across Distinct Biological States."</p>
    <p><i>PLOS Biology</i>, Apr 29;6(4)e96. (2008).</p>
    <br/>

    <p><b>Photocleavage of uracil in the presence of FMN:</b></p>
    <p>P. Burgstaller, T. Hermann, C. Huber, E. Westhof, and M. Famulok, M.</p>
    <p>"Isoalloxazine derivatives promote photocleavage of natural RNAs at G.U base pairs embedded within helices."</p>
    <p><i>Nucleic Acids Ressearch</i>, 25:4018-4027. (1997).</p>
  </body>
</html>
